### PR TITLE
fix: minimize dependencies on 3P libraries

### PR DESCRIPTION
**Pull request by anniefu**
This keeps the functions framework as lightweight as possible, and
avoids pulling in all of the Google Cloud client libraries even though
only one package is needed.

The new `internal/metadata` package is just a copy of
[google-cloud-go/functions/metadata](https://github.com/googleapis/google-cloud-go/tree/main/functions/metadata)

Minimizing what transitive dependencies the Functions Framework has is
important to make them work with buildpacks, where user function code
has to be bundled together with the functions framework. If a function's
dependency conflicts with a Functions Framework dependency and the
function is using vendor, then there can be undefined behavior at build
and runtime.

### CODE REVIEW

**Review by grant — APPROVED** [review]

**Review by matthewrobertson — COMMENTED** [review]
Any idea what we were trying to accomplish by hosting that metadata package in Google APIs? Looks like it was never really finalized: https://github.com/googleapis/google-cloud-go/blob/main/functions/metadata/doc.go#L18

File: internal/events/pubsub/pubsub.go [hunk 1]
```diff
@@ -6,9 +6,8 @@ import (
 	"regexp"
 	"time"
 
-	"cloud.google.com/go/functions/metadata"
```
Comment (matthewrobertson) [inline internal/events/pubsub/pubsub.go:9]: this was a bizarre thing to do...

**Review by matthewrobertson — APPROVED** [review]
LGTM

**Comment by anniefu** [comment]
Not sure what the general intent was, but we use it in Functions Framework to move metadata from the HTTP request body to the golang `Context` object.

In Golang Background Functions, the first parameter `context` ends up encapsulating all the parts of the payload that aren't the data:
```
// Metadata holds Google Cloud Functions metadata.
type Metadata struct {
	// EventID is a unique ID for the event. For example: "70172329041928".
	EventID string `json:"eventId"`
	// Timestamp is the date/time this event was created.
	Timestamp time.Time `json:"timestamp"`
	// EventType is the type of the event. For example: "google.pubsub.topic.publish".
	EventType string `json:"eventType"`
	// Resource is the resource that triggered the event.
	Resource *Resource `json:"resource"`
}
```

That allows us to unmarshal just the data payload as the second parameter of Background Functions, which lets us use reflection to determine the type to unmarshal into.

File: internal/events/pubsub/pubsub.go (hunk 1 above)
Comment (anniefu) [reply internal/events/pubsub/pubsub.go:9]: was it? I was just making structs out of structs 🤔

**Pull request merged**